Reload tools if a macro changes

mvdbeek opened the pull request:
We do this by recording any macro that is required to load a tool.  When
loading a tool we register the macro to be watched, and if the macro changes we
reload all corresponding tools.

xref: https://github.com/galaxyproject/galaxy/issues/3987

mvdbeek committed:
Reload tools if a macro changes

We do this by recording any macros that are required to load a tool.  When
loading a tool we register the macro to be watched, and if the macro changes we
reload all corresponding tools.

jmchilton commented:
Working on a ~PR~ commit to preserve some backward compatibility with this - this doesn't break anything in Galaxy but it will break Planemo.

jmchilton committed:
Rework #4635 to preserve backward compatibility on modules shared via galaxy-lib.

Also cleanup comments and such in xml_macros to reflect this doesn't apply just to tools.

jmchilton commented:
@mvdbeek does d3d134030aded5b29e62c4c1d5946f3266a34827 look okay to you?

mvdbeek commented:
much better, thanks a lot @jmchilton !

jmchilton commented:
I'm already super excited for 18.01 - this is very cool @mvdbeek - thanks!

natefoo commented on lib/galaxy/tools/toolbox/base.py:
```diff
@@ -745,10 +747,13 @@ def load_tool(self, config_file, guid=None, repository_id=None, use_cached=False
+            if self._tool_watcher:
+                self._tool_watcher.watch_file(config_file, tool.id)
+            if self._tool_config_watcher:
+                [self._tool_config_watcher.watch_file(macro_path) for macro_path in tool._macro_paths]
```
Just something I ran in to while looking at another issue today - I was confused as to why a macro was part of the `ToolConfWatcher` (the one that watches tool *panel* configs e.g. `shed_tool_conf.xml`) since I don't have tool watching enabled, and it turns out this was the reason. Presumably this is because the tool watcher only reloads single tools, so this was the quickest way to get macro watching enabled.

Maybe there should be macro watcher, or the tool watcher should register tool IDs with macros so it can reload them instead of using the toolbox reloader? Not a big deal, but a little bit of an unintentional side effect that macros are watched even when tool reloading is off. I guess we could also just case this so the macros aren't added to the watcher if tool watching is not enabled.
